File: src/api-service/__app__/onefuzzlib/request_access.py
from typing import Dict, List, Optional
from uuid import UUID

from onefuzztypes.models import ApiAccessRule


class RuleConflictError(Exception):
    def __init__(self, message: str) -> None:
        super(RuleConflictError, self).__init__(message)
        self.message = message


class RequestAccess:
    """
    Stores the rules associated with a the request paths
    """

    class Rules:
        allowed_groups_ids: List[UUID]

        def __init__(self, allowed_groups_ids: List[UUID] = []) -> None:
            self.allowed_groups_ids = allowed_groups_ids

    class Node:
        # http method -> rules
        rules: Dict[str, "RequestAccess.Rules"]
        # path -> node
        children: Dict[str, "RequestAccess.Node"]

        def __init__(self) -> None:
            self.rules = {}
            self.children = {}
            pass

    root: Node

    def __init__(self) -> None:
        self.root = RequestAccess.Node()

    def __add_url__(self, methods: List[str], path: str, rules: Rules) -> None:
        methods = list(map(lambda m: m.upper(), methods))

        segments = [s for s in path.split("/") if s != ""]
        if len(segments) == 0:
            return

        current_node = self.root
        current_segment_index = 0

        while current_segment_index < len(segments):
            current_segment = segments[current_segment_index]
            if current_segment in current_node.children:
                current_node = current_node.children[current_segment]
                current_segment_index = current_segment_index + 1
            else:
                break
        # we found a node matching this exact path
        # This means that there is an existing rule causing a conflict
        if current_segment_index == len(segments):
            for method in methods:
                if method in current_node.rules:
                    raise RuleConflictError(f"Conflicting rules on {method} {path}")

        while current_segment_index < len(segments):
            current_segment = segments[current_segment_index]
            current_node.children[current_segment] = RequestAccess.Node()
            current_node = current_node.children[current_segment]
            current_segment_index = current_segment_index + 1

        for method in methods:
            current_node.rules[method] = rules

    def get_matching_rules(self, method: str, path: str) -> Optional[Rules]:
        method = method.upper()
        segments = [s for s in path.split("/") if s != ""]
        current_node = self.root
        current_rule = None

        if method in current_node.rules:
            current_rule = current_node.rules[method]

        current_segment_index = 0

        while current_segment_index < len(segments):
            current_segment = segments[current_segment_index]
            if current_segment in current_node.children:
                current_node = current_node.children[current_segment]
            elif "*" in current_node.children:
                current_node = current_node.children["*"]
            else:
                break

            if method in current_node.rules:
                current_rule = current_node.rules[method]
            current_segment_index = current_segment_index + 1
        return current_rule

    @classmethod
    def build(cls, rules: Dict[str, ApiAccessRule]) -> "RequestAccess":
        request_access = RequestAccess()
        for endpoint in rules:
            rule = rules[endpoint]
            request_access.__add_url__(
                rule.methods,
                endpoint,
                RequestAccess.Rules(allowed_groups_ids=rule.allowed_groups),
            )

        return request_access

File: src/ApiService/ApiService/onefuzzlib/RequestAccess.cs

using System.Net.Http;

namespace Microsoft.OneFuzz.Service;

public class RequestAccess {
    private readonly Node _root = new();

    public record Rules(IReadOnlyList<Guid> AllowedGroupsIds);
    sealed record Node(
        // HTTP Method -> Rules
        Dictionary<HttpMethod, Rules> Rules,
        // Path Segment -> Node
        Dictionary<string, Node> Children) {
        public Node() : this(new(), new()) { }
    }

    private void AddUri(IEnumerable<HttpMethod> methods, string path, Rules rules) {
        var segments = path.Split('/', StringSplitOptions.RemoveEmptyEntries);
        if (!segments.Any()) {
            return;
        }

        var currentNode = _root;
        var currentSegmentIndex = 0;

        while (currentSegmentIndex < segments.Length) {
            var currentSegment = segments[currentSegmentIndex];
            if (currentNode.Children.ContainsKey(currentSegment)) {
                currentNode = currentNode.Children[currentSegment];
                currentSegmentIndex++;
            } else {
                break;
            }
        }

        // we found a node matching this exact path
        // This means that there is an existing rule causing a conflict
        if (currentSegmentIndex == segments.Length) {
            var conflictingMethod = methods.FirstOrDefault(m => currentNode.Rules.ContainsKey(m));
            if (conflictingMethod is not null) {
                throw new RuleConflictException($"Conflicting rules on {conflictingMethod} {path}");
            }
        }

        while (currentSegmentIndex < segments.Length) {
            var currentSegment = segments[currentSegmentIndex];
            currentNode = currentNode.Children[currentSegment] = new Node();
            currentSegmentIndex++;
        }

        foreach (var method in methods) {
            currentNode.Rules[method] = rules;
        }
    }

    public static RequestAccess Build(IDictionary<string, ApiAccessRule> rules) {
        var result = new RequestAccess();
        foreach (var (endpoint, rule) in rules) {
            result.AddUri(rule.Methods.Select(x => new HttpMethod(x)), endpoint, new Rules(rule.AllowedGroups));
        }

        return result;
    }

    public Rules? GetMatchingRules(HttpMethod method, string path) {
        var segments = path.Split("/", StringSplitOptions.RemoveEmptyEntries);

        var currentNode = _root;
        _ = currentNode.Rules.TryGetValue(method, out var currentRule);

        foreach (var currentSegment in segments) {
            if (currentNode.Children.TryGetValue(currentSegment, out var node)) {
                currentNode = node;
            } else if (currentNode.Children.TryGetValue("*", out var starNode)) {
                currentNode = starNode;
            } else {
                break;
            }

            if (currentNode.Rules.TryGetValue(method, out var rule)) {
                currentRule = rule;
            }
        }

        return currentRule;
    }
}

public sealed class RuleConflictException : Exception {
    public RuleConflictException(string? message) : base(message) {
    }
}
